storage/pebble/blocks.go

```diff
@@ -0,0 +1,180 @@
+package pebble
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/onflow/flow-evm-gateway/storage"
+	errs "github.com/onflow/flow-evm-gateway/storage/errors"
+	"github.com/onflow/flow-go/fvm/evm/types"
+)
+
+var _ storage.BlockIndexer = &Blocks{}
+
+type BlockOption func(block *Blocks) error
+
+// WithInitHeight sets the first and last height to the provided value,
+// this should be used to initialize an empty database, if the first and last
+// heights are already set an error will be returned.
+func WithInitHeight(height uint64) BlockOption {
+	return func(block *Blocks) error {
+		return block.storeInitHeight(height)
+	}
+}
+
+type Blocks struct {
+	store       *Storage
+	mux         sync.RWMutex
+	heightCache map[byte]uint64
+}
+
+func NewBlocks(store *Storage, opts ...BlockOption) (*Blocks, error) {
+	blk := &Blocks{
+		store:       store,
+		mux:         sync.RWMutex{},
+		heightCache: make(map[byte]uint64),
+	}
+
+	for _, opt := range opts {
+		if err := opt(blk); err != nil {
+			return nil, err
+		}
+	}
+
+	return blk, nil
+}
+
+func (b *Blocks) Store(block *types.Block) error {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	val, err := block.ToBytes()
+	if err != nil {
+		return err
+	}
+
+	id, err := block.Hash()
+	if err != nil {
+		return err
+	}
+
+	// todo batch operations
+	height := uint64Bytes(block.Height)
+	if err := b.store.set(blockHeightKey, height, val); err != nil {
+		return err
+	}
+
+	// todo check if what is more often used block by id or block by height and fix accordingly if needed
```

Comment:
I believe that block by id is probably more widely used. There are JSON-RPC endpoints for both versions, but the block number is probably easier to use.

Reply:
I will just check logs after some time to see what is more often requested.